kubeadm: Support skipping addons image pull

ruquanzhao opened the pull request:
Add field `AddOns` in ClusterConfiguration to save status for `join` and `upgrade` to read.

Signed-off-by: Ruquan Zhao <ruquan.zhao@arm.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubeadm/issues/2603

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

neolit123 commented:
> Add field AddOns in ClusterConfiguration to save status for join and upgrade to read.

all kubeadm api changes need thorough discussion.
also v1beta3 is locked and new features cannot be added.

i don't remember what alternatives were discussed on the k/kubeadm ticket.

pacoxu commented:
/cc @chendave 
/kind feature
/release-note-edit ```release-note
kubeadm: skip image pull for addons that are skipped
```
/ok-to-test
/priority important-longterm
/triage accepted

pacoxu commented:
> all kubeadm api changes need thorough discussion.
> also v1beta3 is locked and new features cannot be added.

/hold 
sorry for not reading this comment.

chendave commented:
/assign

chendave commented:
/remove-lifecycle stale

chendave commented:
@neolit123  sorry for ping again!

Is there any plan for v1beta4? there are several PRs that is labeled for v1beta4 and is pending for quite a while.

neolit123 commented:
> @neolit123 sorry for ping again!
> 
> Is there any plan for v1beta4? there are several PRs that is labeled for v1beta4 and is pending for quite a while.

as we discussed earlier, changes get batched per new api version. if you look at k/kubeadm, there are a number of api-change labeled tickets.

we need to sit down and plan what can we release in v1beta4 and who can work on it. you can drive this plan if you'd like for .29. create an issue in kubeadm called v1beta4, link all tickets and let's discuss.

chendave commented:
@ruquanzhao can you pls rebase this on top of v1beta4?

ruquanzhao commented:
/retest

ruquanzhao commented:
/retest

chendave commented on cmd/kubeadm/app/images/images_test.go:
```diff
@@ -217,15 +218,47 @@ func TestGetAllImages(t *testing.T) {
 			},
 			expect: constants.Etcd,
 		},
+		// add-ons' images should follow ClusterConfiguration.AddOns
```
```suggestion
		// add-on images returned should respect the configuration of ClusterConfiguration.AddOns
```

chendave commented on cmd/kubeadm/app/images/images_test.go:
```diff
@@ -217,15 +218,47 @@ func TestGetAllImages(t *testing.T) {
+			},
+			expect:     "",
+			expectNots: []string{constants.CoreDNSImageName, constants.KubeProxy},
+		},
```
Add one testcase here, where both or either of the addon is not configured, but should find them in the image list.

ruquanzhao replied:
Done.

chendave commented on cmd/kubeadm/app/images/images_test.go:
```diff
@@ -217,15 +218,47 @@ func TestGetAllImages(t *testing.T) {
+					"kube-proxy": false,
+				},
+			},
+			expect:     "",
```
```suggestion

```

It's an empty string by default.

ruquanzhao replied:
Done.

chendave commented on cmd/kubeadm/app/images/images_test.go:
```diff
@@ -217,15 +218,47 @@ func TestGetAllImages(t *testing.T) {
+			}
 			for _, img := range imgs {
 				if strings.Contains(img, tc.expect) {
 					return
```
those logic should be updated, `strings.Contains(img, "")` will already return true.

ruquanzhao replied:
Thanks ! Done.

chendave reviewed (commented):
Overall LGTM.

Need review or feedback from others @neolit123 @SataQiu @pacoxu

chendave commented:
need rebase as well, but you hold until there is more feedback from other reviewers.

chendave commented:
pls update the this file: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go#L37

to have this included in the v1beta4 release note.

/release-note-none

chendave commented:
@ruquanzhao Can you rebase again to remove this commit: [Merge branch 'master' into skipaddons](https://github.com/kubernetes/kubernetes/pull/114534/commits/2ac07b56607bce08a64354166569ca24007a30aa) ?

chendave commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -34,7 +34,9 @@ limitations under the License.
 //   - Add `ClusterConfiguration.EncryptionAlgorithm` that can be used to set the asymmetric encryption algorithm
 //     used for this cluster's keys and certificates. Can be "RSA" (default algorithm, key size is 2048) or
-//     "ECDSA" (uses the P-256 eliptic curve).
+//     "ECDSA" (uses the P-256 elliptic curve).
```
revert this change pls.

chendave commented:
- update the master first
- rebase your change on the top of master (your change has the latest update on master)
there is still merge conflict in this pr.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -163,6 +163,17 @@ func newCmdInit(out io.Writer, initOptions *initOptions) *cobra.Command {
 		if len(initRunner.Options.SkipPhases) == 0 {
 			initRunner.Options.SkipPhases = data.cfg.SkipPhases
 		}
+		// dump AddOns status by skip-phases
+		if isStringInSlice("addon/coredns", initRunner.Options.SkipPhases) {
+			data.cfg.AddOns["coredns"] = false
+		}
+		if isStringInSlice("addon/kube-proxy", initRunner.Options.SkipPhases) {
+			data.cfg.AddOns["kube-proxy"] = false
+		}
+		if isStringInSlice("addon", initRunner.Options.SkipPhases) {
+			data.cfg.AddOns["coredns"] = false
+			data.cfg.AddOns["kube-proxy"] = false
+		}
```
if the user deploys all addons by just calling `kubeadm init` without skipping phases, does it mean we have to set this Addons map key/value to `"someaddonname": true`?

chendave replied:
> if the user deploys all addons by just calling kubeadm init without skipping phases, does it mean we have to set this Addons map key/value to "someaddonname": true

By default, without the set in the config file, the addon will be pulled anyway, current logic is the skip will only happen if it's set to false explicitly in the config file. @ruquanzhao is that right?

ruquanzhao replied:
Thanks @chendave for explanation. 
In this implement, `skip-phases` will set `addons` false, and the `addons` only take effect when set to false(by `skip-phases` or user set in configfile)

It does seem more reasonable to use some field like `SkippedAddons`.

SataQiu replied:
+1 for `SkippedAddons`

neolit123 commented on cmd/kubeadm/app/images/images.go:
```diff
@@ -104,8 +103,16 @@ func GetControlPlaneImages(cfg *kubeadmapi.ClusterConfiguration) []string {
-	images = append(images, GetDNSImage(cfg))
+	if flag, exist := cfg.AddOns["coredns"]; !exist || flag {
+		images = append(images, GetDNSImage(cfg))
+	}
```
do we need to tell the user with a high verbosity `klog.V(x).Infof` that the pull of certain addons images are skipped?

ruquanzhao replied:
Done.

chendave commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -35,6 +35,8 @@ limitations under the License.
 //     used for this cluster's keys and certificates. Can be "RSA" (default algorithm, key size is 2048) or
 //     "ECDSA" (uses the P-256 elliptic curve).
+//   - Add `ClusterConfiguration.AddOns` that can be used to skip pulling add-on images(coredns and kube-proxy)
+//     if explicitly configured as `false` or addon phases are skipped.
```
```suggestion
//     If explicitly configured as `false`, pull image for add-on will be skipped in preflight phase
```

neolit123 reviewed (commented):
thanks for the updates.
added some comments / questions / discussion topics.
i think we also need to call the Cluster API foks for review, because they manage the addons with custom logic.

/hold

chendave reviewed (commented):
/approve
/hold

/assign @neolit123 @pacoxu @SataQiu for a double check.

ruquanzhao commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -143,6 +143,9 @@ type ClusterConfiguration struct {
 	// Can be "RSA" (default algorithm, key size is 2048) or "ECDSA" (uses the P-256 elliptic curve).
 	EncryptionAlgorithm EncryptionAlgorithmType
+
+	// AddOns status
+	AddOns map[string]bool
```
Agree with that. If user use `skip-phases=addon` but not set `SkippedAddons`,  `kubeadm` will help with set `SkippedAddons` true, is that reasonable?

neolit123 commented:
spoke with `@fabriziopandini` about this today.
his feedback is the following.

we already have a DNS under ClusterConfiguration.
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go#L182-L189

kubeadm will probably continue to deploy a DNS and proxy addons by default, so we can have some extra logic in ClusterCnfiguraiton and follow a better-defined declarative approach in v1beta4:
- DNS.Disabled = true/false (bool field)
- Proxy.Disabled = true/false (bool field)

(both are false by default and don't need defaulting)

`Proxy` is a new struct.
if the user skips the addon/proxy or addon/dns phases of `kubeadm init` we can set these bool fields accordingly.

side tasks:
- we also need to remove `DNSAddOnType` from v1beta4 as it's not used anymore.
- decide what to do with this upgrade code:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/postupgrade.go#L144-L192
if we now have a way to track skipped addons should we not check for config-map existence.

cc @SataQiu @pacoxu 
for comments as well.

pacoxu commented:
For upgrade, does `DNS.Disabled` mean removing the DNS-related resource? Or does `DNS.Disabled` mean skipping DNS-related resource updates?

neolit123 commented:
> For upgrade, does `DNS.Disabled` mean removing the DNS-related resource? Or does `DNS.Disabled` mean skipping DNS-related resource updates?

we should avoid mutations during upgrade, so it would mean skipping.

SataQiu commented:
> decide what to do with this upgrade code:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/postupgrade.go#L144-L192
if we now have a way to track skipped addons should we not check for config-map existence?

During the migration period(v1beta3 -> v1beta4), I think we still need this check. 
But eventually, I think we can change the way we handle this check. Maybe throws an error when ClusterConfiguration indicates the DNS upgrade should be done but the config map can not be found.

Commit:
Support skipping pulling addons image.

Add field DNS.Disabled and Proxy.Disabled in ClusterConfiguration to save status for `join` to read.
Signed-off-by: Ruquan Zhao <ruquan.zhao@arm.com>

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -177,6 +180,15 @@ type APIServer struct {
 	// ImageMeta allows to customize the image used for the DNS component
 	ImageMeta `json:",inline"`
+
+	// Disabled specifies whether to disable DNS addon in the cluster
```
```suggestion
	// Disabled specifies whether to disable this addon in the cluster
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -177,6 +180,15 @@ type APIServer struct {
+	Disabled bool
+}
+
+// Proxy defines the Proxy addon that should be used in the cluster
```
```suggestion
// Proxy defines the proxy addon that should be used in the cluster
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -177,6 +180,15 @@ type APIServer struct {
+
+// Proxy defines the Proxy addon that should be used in the cluster
+type Proxy struct {
+	// Disabled specifies whether to disable  addon in the cluster
```
```suggestion
	// Disabled specifies whether to disable this addon in the cluster
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -177,6 +180,15 @@ type APIServer struct {
 type DNS struct {
 	// ImageMeta allows to customize the image used for the DNS component
```
```suggestion
	// ImageMeta allows to customize the image used for the DNS addon
```

neolit123 replied:
this is an existing comment, but please change it.

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -35,6 +35,8 @@ limitations under the License.
 //     used for this cluster's keys and certificates. Can be "RSA" (default algorithm, key size is 2048) or
 //     "ECDSA" (uses the P-256 elliptic curve).
+//   - Add `ClusterConfiguration.DNS.Disabled` and `ClusterConfiguration.Proxy.Disabled` that can be used to disable
+//     CoreDNS and kube-proxy setup when set to true.
```
```suggestion
//     the CoreDNS and kube-proxy addons during cluster initialization. Skipping the related addons phases,
//     during cluster creation will set the same fields to `false`.
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go:
```diff
@@ -183,6 +186,15 @@ type APIServer struct {
 	// ImageMeta allows to customize the image used for the DNS component
 	ImageMeta `json:",inline"`
+
+	// Disabled specifies whether to disable DNS addon in the cluster
```
```suggestion
	// Disabled specifies whether to disable this addon in the cluster
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go:
```diff
@@ -183,6 +186,15 @@ type APIServer struct {
+
+// Proxy defines the Proxy addon that should be used in the cluster
+type Proxy struct {
+	// Disabled specifies whether to disable  addon in the cluster
```
```suggestion
	// Disabled specifies whether to disable this addon in the cluster
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -163,6 +163,24 @@ func newCmdInit(out io.Writer, initOptions *initOptions) *cobra.Command {
 		if len(initRunner.Options.SkipPhases) == 0 {
 			initRunner.Options.SkipPhases = data.cfg.SkipPhases
 		}
+		// dump DNS and Proxy status by skip-phases
```
```suggestion
		// Update the Proxy and DNS "Disabled" status based on skipped addon phases
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -163,6 +163,24 @@ func newCmdInit(out io.Writer, initOptions *initOptions) *cobra.Command {
+			data.cfg.DNS.Disabled = true
+			data.cfg.Proxy.Disabled = true
+		}
+		// If DNS or Proxy is disabled, skip the corresponding phase.
+		if data.cfg.DNS.Disabled {
+			initRunner.Options.SkipPhases = append(initRunner.Options.SkipPhases, "addon/coredns")
+		}
+		if data.cfg.Proxy.Disabled {
+			initRunner.Options.SkipPhases = append(initRunner.Options.SkipPhases, "addon/kube-proxy")
+		}
```
this will cause a double append - having the `addon/* `phases in `SkipPhases` two times?
this logic should be after:
```
if len(initRunner.Options.SkipPhases) == 0 {
	initRunner.Options.SkipPhases = data.cfg.SkipPhases
}
```
before the logic to update DIsabled fields?

chendave replied:
+1, this suggestion by @neolit123 is that if it's set by the CLI flags, then just use it otherwise read if from cfg.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -163,6 +163,24 @@ func newCmdInit(out io.Writer, initOptions *initOptions) *cobra.Command {
+			data.cfg.DNS.Disabled = true
+			data.cfg.Proxy.Disabled = true
+		}
+		// If DNS or Proxy is disabled, skip the corresponding phase.
```
```suggestion
		// If the DNS or Proxy addons are disabled, skip the corresponding phase
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/images/images.go:
```diff
@@ -104,8 +103,19 @@ func GetControlPlaneImages(cfg *kubeadmapi.ClusterConfiguration) []string {
-	images = append(images, GetDNSImage(cfg))
+	// if DNS addon is not disable then add the image
+	if cfg.DNS.Disabled {
+		klog.V(1).Infof("CoreDNS image pull is skipped")
```
```suggestion
		klog.V(1).Infof("skipping the CoreDNS image pull since the bundled addon is disabled")
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/images/images.go:
```diff
@@ -104,8 +103,19 @@ func GetControlPlaneImages(cfg *kubeadmapi.ClusterConfiguration) []string {
+
+	// if Proxy addon is not disable then add the image
+	if cfg.Proxy.Disabled {
+		klog.V(1).Infof("kube-proxy image pull is skipped")
```
```suggestion
		klog.V(1).Infof("skipping the kube-proxy image pull since the bundled addon is disabled")
```

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -163,6 +163,24 @@ func newCmdInit(out io.Writer, initOptions *initOptions) *cobra.Command {
 		if len(initRunner.Options.SkipPhases) == 0 {
 			initRunner.Options.SkipPhases = data.cfg.SkipPhases
 		}
+		// dump DNS and Proxy status by skip-phases
+		if isStringInSlice("addon/coredns", initRunner.Options.SkipPhases) {
+			data.cfg.DNS.Disabled = true
+		}
+		if isStringInSlice("addon/kube-proxy", initRunner.Options.SkipPhases) {
+			data.cfg.Proxy.Disabled = true
+		}
+		if isStringInSlice("addon", initRunner.Options.SkipPhases) {
+			data.cfg.DNS.Disabled = true
+			data.cfg.Proxy.Disabled = true
+		}
+		// If DNS or Proxy is disabled, skip the corresponding phase.
+		if data.cfg.DNS.Disabled {
+			initRunner.Options.SkipPhases = append(initRunner.Options.SkipPhases, "addon/coredns")
+		}
+		if data.cfg.Proxy.Disabled {
+			initRunner.Options.SkipPhases = append(initRunner.Options.SkipPhases, "addon/kube-proxy")
+		}
```
all these nwe changes in SetDataInitializer should be moved to a new function - e.g.
```
manageSkippedAddons(cfg *ClusterConfiguration, skipPhases []string) 
```

and it should be unit tested.

ruquanzhao replied:
Done. Moved it to a new function `manageSkippedAddons` and added some unittest.

neolit123 commented on cmd/kubeadm/app/images/images_test.go:
```diff
@@ -215,23 +229,63 @@ func TestGetAllImages(t *testing.T) {
+			},
+		},
+		{
+			name: "CoreDNS and kube-proxy image is returned",
```
```suggestion
			name: "CoreDNS and kube-proxy images are returned",
```

ruquanzhao replied:
Done.

chendave commented on cmd/kubeadm/app/images/images_test.go:
```diff
@@ -215,23 +229,63 @@ func TestGetAllImages(t *testing.T) {
+		},
+		{
+			name: "CoreDNS and kube-proxy image is returned",
+			cfg:  &kubeadmapi.ClusterConfiguration{},
```
add one testcase to have `Disabled` set to `false` for the two addon and get back them in the list of `expectedImages`.

ruquanzhao replied:
Done.

chendave commented:
@ruquanzhao pls fix @neolit123 's comments and unit job, then this is good to go.

NOTE:  only a few days before the code freeze.

ruquanzhao commented:
I tried this on my machine, and it works as expected.
Cluster.config
```
apiVersion: kubeadm.k8s.io/v1beta4
kind: ClusterConfiguration
kubernetesVersion: 1.28.2
dns:
  disabled: true
proxy:
  disabled: true
```

`kubectl get nodes -A`
<img width="717" alt="Screenshot 2023-10-25 at 16 27 10" src="https://github.com/kubernetes/kubernetes/assets/49401013/a7cbdafe-1c88-41dd-b9e9-292c71dba80a">
`kubectl get cm -n   kube-system   kubeadm-config  -o yaml`
<img width="650" alt="Screenshot 2023-10-25 at 16 27 54" src="https://github.com/kubernetes/kubernetes/assets/49401013/0f0a2bcb-9f45-40b4-95b9-6f5cc376276e">

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -527,3 +529,38 @@ func (d *initData) PatchesDir() string {
+
+// manageSkippedAddons syncs proxy and DNS "Disabled" status and skipPhases.
+func manageSkippedAddons(cfg *kubeadmapi.InitConfiguration, skipPhases []string) []string {
+	corednsPhase := "addon/coredns"
+	proxyPhase := "addon/kube-proxy"
+	addonPhase := "addon"
```
move these into a `const (` ... block

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -527,3 +529,38 @@ func (d *initData) PatchesDir() string {
 }
+
+// manageSkippedAddons syncs proxy and DNS "Disabled" status and skipPhases.
+func manageSkippedAddons(cfg *kubeadmapi.InitConfiguration, skipPhases []string) []string {
```
```
// If DNS is disabled, append the phase
...
// If proxy is disabled, append the phase
...
// if a DNS skip phase is present update the disabled field
...
// if a proxy skip phase is present update the disabled field
...
// if whole addon phase is skipped update both DNS and proxy Disable fields
...

```

the logic can be changed to this

neolit123 replied:
```suggestion
func manageSkippedAddons(cfg *kubeadmapi.ClusterConfiguration, skipPhases []string) []string {
```

neolit123 commented on cmd/kubeadm/app/cmd/init_test.go:
```diff
@@ -220,6 +221,154 @@ func TestNewInitData(t *testing.T) {
 }
 
+func TestManageSkippedAddons(t *testing.T) {
+	// sync skipPhases to Cfg
+	// sync cfg to skipphases
+	// if disabled=false and skipphaase is true, use skipphases.
```
```suggestion
```

comment can be removed

neolit123 commented on cmd/kubeadm/app/cmd/init_test.go:
```diff
@@ -220,6 +221,154 @@ func TestNewInitData(t *testing.T) {
+		expectedSkipPhases []string
+	}{
+		{
+			name: "disable proxy and DNS if 'addon' in skipPhases",
```
```suggestion
			name: "disable proxy and DNS if 'addon' is in skipPhases",
```

neolit123 commented on cmd/kubeadm/app/cmd/init_test.go:
```diff
@@ -220,6 +221,154 @@ func TestNewInitData(t *testing.T) {
+			expectedSkipPhases: []string{"addon"},
+		},
+		{
+			name: "disable proxy and DNS if 'addon/coredns' and 'addon/kube-proxy' in skipPhases",
```
```suggestion
			name: "disable proxy and DNS if 'addon/coredns' and 'addon/kube-proxy' are in skipPhases",
```

neolit123 commented on cmd/kubeadm/app/cmd/init_test.go:
```diff
@@ -220,6 +221,154 @@ func TestNewInitData(t *testing.T) {
+			expectedSkipPhases: []string{"addon/coredns", "addon/kube-proxy"},
+		},
+		{
+			name: "disable proxy if 'addon/kube-proxy' in skipPhases",
```
```suggestion
			name: "disable proxy if 'addon/kube-proxy' is in skipPhases",
```

neolit123 commented on cmd/kubeadm/app/cmd/init_test.go:
```diff
@@ -220,6 +221,154 @@ func TestNewInitData(t *testing.T) {
+			expectedSkipPhases: []string{"addon/kube-proxy"},
+		},
+		{
+			name: "disable DNS if 'addon/coredns' in skipPhases",
```
```suggestion
			name: "disable DNS if 'addon/coredns' is in skipPhases",
```

neolit123 commented on cmd/kubeadm/app/cmd/init_test.go:
```diff
@@ -220,6 +221,154 @@ func TestNewInitData(t *testing.T) {
+			expectedSkipPhases: []string{"addon/coredns"},
+		},
+		{
+			name: "add 'addon/coredns' and 'addon/kube-proxy' to skipPhases if DNS and proxy is disabled",
```
```suggestion
			name: "add 'addon/coredns' and 'addon/kube-proxy' to skipPhases if DNS and proxy are disabled",
```

neolit123 commented on cmd/kubeadm/app/cmd/init_test.go:
```diff
@@ -220,6 +221,154 @@ func TestNewInitData(t *testing.T) {
+			expectedSkipPhases: []string{"addon/coredns", "addon/kube-proxy"},
+		},
+		{
+			name: "don't add duplicated skipPhase in skipPhases",
```
```suggestion
			name: "don't add duplicates in skipPhases",
```

neolit123 reviewed (commented):
added some further comments about improving the logic and fixing some typos in unit test names. thanks

neolit123 commented:
in the PR description, please change:
```
Fixes #https://github.com/kubernetes/kubeadm/issues/2603
```

to:
```
Fixes ttps:h//github.com/kubernetes/kubeadm/issues/2603
```

neolit123 commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -431,6 +431,15 @@ const (
 	EtcdUserName string = "kubeadm-etcd"
 	// ServiceAccountKeyReadersGroupName is the group of users that are allowed to read the service account private key.
 	ServiceAccountKeyReadersGroupName string = "kubeadm-sa-key-readers"
+
+	// CoreDNSPhase is the name of CoreDNS phase in "kubeadm init"
+	CoreDNSPhase string = "addon/coredns"
+
+	// KubeProxyPhase is the name of kube-proxy phase during "kubeadm init"
+	KubeProxyPhase string = "addon/kube-proxy"
+
+	// AddonPhase is the name of addon phase during "kubeadm init"
+	AddonPhase string = "addon"
```
please move these to init.go where they are used above the manageSkipAddons function, lowercase them as well (make them local)

ruquanzhao replied:
Done.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -527,3 +529,38 @@ func (d *initData) PatchesDir() string {
+// manageSkippedAddons syncs proxy and DNS "Disabled" status and skipPhases.
+func manageSkippedAddons(cfg *kubeadmapi.ClusterConfiguration, skipPhases []string) []string {
+
+	// If the DNS or Proxy addons are disabled, skip the corresponding phase
+	if cfg.DNS.Disabled && !isPhaseInSkipPhases(kubeadmconstants.AddonPhase, skipPhases) && !isPhaseInSkipPhases(kubeadmconstants.CoreDNSPhase, skipPhases) {
+		skipPhases = append(skipPhases, kubeadmconstants.CoreDNSPhase)
+	}
+	if cfg.Proxy.Disabled && !isPhaseInSkipPhases(kubeadmconstants.AddonPhase, skipPhases) && !isPhaseInSkipPhases(kubeadmconstants.KubeProxyPhase, skipPhases) {
+		skipPhases = append(skipPhases, kubeadmconstants.KubeProxyPhase)
+	}
```
```suggestion
	// If the DNS or Proxy addons are disabled, skip the corresponding phase.
	// Alternatively, update the proxy and DNS "Disabled" status based on skipped addon phases.
	var (
		skipDNSPhase = false
		skipProxyPhase = false
	)
	if isPhaseInSkipPhases(kubeadmconstants.AddonPhase, skipPhases) {
		skipDNSPhase = true
		skipProxyPhase = true
		cfg.DNS.Disabled = true
		cfg.Proxy.Disabled = true
	}
	if isPhaseInSkipPhases(kubeadmconstants.CoreDNSPhase, skipPhases) {
		skipDNSPhase = true
		cfg.DNS.Disabled = true
	}
	if isPhaseInSkipPhases(kubeadmconstants.KubeProxyPhase, skipPhases) {
		skipProxyPhase = true
		cfg.Proxy.Disabled = true
	}
	if cfg.DNS.Disabled && !skipDNSPhase {
			skipPhases = append(skipPhases, kubeadmconstants.CoreDNSPhase)
	}
	if cfg.DNS.Disabled && !skipProxyPhase {
			skipPhases = append(skipPhases, kubeadmconstants.KubeProxyPhase)
	}
```

can the whole function be changed to this?

ruquanzhao replied:
Yes, of course. You saved us from checking it again and again.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go:
```diff
@@ -92,6 +92,10 @@ func Convert_kubeadm_NodeRegistrationOptions_To_v1beta3_NodeRegistrationOptions(
 	return autoConvert_kubeadm_NodeRegistrationOptions_To_v1beta3_NodeRegistrationOptions(in, out, s)
 }
 
+func Convert_kubeadm_DNS_To_v1beta3_DNS(in *kubeadm.DNS, out *DNS, s conversion.Scope) error {
```
```suggestion
// Convert_kubeadm_DNS_To_v1beta3_DNS converts a private DNS to public DNS.
func Convert_kubeadm_DNS_To_v1beta3_DNS(in *kubeadm.DNS, out *DNS, s conversion.Scope) error {
```

neolit123 replied:
xref https://github.com/kubernetes/kubernetes/pull/121511

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go:
```diff
@@ -181,8 +184,19 @@ type APIServer struct {
+type Proxy struct {
+	// Disabled specifies whether to disable this addon in the cluster
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
```
@chendave do you think we can face a problem where we need this to be a *bool?
i.e. we must know when the user set it explicitly to `false`?
(same for proxy)

i think, no.

chendave replied:
By default, it's false. There is a conflict if user manually specified it's as false, but the phase is intending to skip via the flags `--skip-phases`, I think the logic is already covered by `manageSkippedAddons`, so I am fine with it.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -66,6 +66,17 @@ type initOptions struct {
 }
 
+const (
+	// CoreDNSPhase is the name of CoreDNS phase in "kubeadm init"
+	coreDNSPhase string = "addon/coredns"
```
```suggestion
	// CoreDNSPhase is the name of CoreDNS sub-phase in "kubeadm init"
	coreDNSPhase = "addon/coredns"
```

ruquanzhao replied:
done

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -66,6 +66,17 @@ type initOptions struct {
+	// CoreDNSPhase is the name of CoreDNS phase in "kubeadm init"
+	coreDNSPhase string = "addon/coredns"
+
+	// KubeProxyPhase is the name of kube-proxy phase during "kubeadm init"
+	kubeProxyPhase string = "addon/kube-proxy"
```
```suggestion
	// KubeProxyPhase is the name of kube-proxy subphase during "kubeadm init"
	kubeProxyPhase = "addon/kube-proxy"
```

ruquanzhao replied:
done

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -66,6 +66,17 @@ type initOptions struct {
+	kubeProxyPhase string = "addon/kube-proxy"
+
+	// AddonPhase is the name of addon phase during "kubeadm init"
+	addonPhase string = "addon"
```
```suggestion
	addonPhase = "addon"
```

ruquanzhao replied:
done

neolit123 reviewed (commented):
/approve

added a few minor comments, but this is looking good to go.

chendave commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -35,6 +35,9 @@ limitations under the License.
 //     "ECDSA" (uses the P-256 elliptic curve).
+//   - Add `ClusterConfiguration.DNS.Disabled` and `ClusterConfiguration.Proxy.Disabled` that can be used to disable
+//     the CoreDNS and kube-proxy addons during cluster initialization. Skipping the related addons phases,
+//     during cluster creation will set the same fields to `false`.
```
note to myself: we need to update the website once v1beta4 is open.

neolit123 replied:
yep, there is authored content we need to update. the api godocs are auto-generated though.

chendave commented:
I think all comments from @neolit123 has been addressed, so this is LGTM.

/lgtm

Happy to see another big feature is going to be landed for v1beta4, good job @ruquanzhao !

Let's hold for a few hours in case @pacoxu or @SataQiu wants to take a final review, otherwise I am going to unhold this before the end of the day!

chendave commented:
@neolit123 I am thinking open v1beta4 for a pilot use, so that we can fix some bugs if there is any before the official release of v1beta4.

It's now hard-coding as false, so that v1beta4 is not possible for `kubeadm init`
https://github.com/kubernetes/kubernetes/blob/c8125c40299513470c7b992169233edf086755da/cmd/kubeadm/app/util/config/initconfiguration.go#L295

WDYT?

neolit123 commented:
> @neolit123 I am thinking open v1beta4 for a pilot use, so that we can fix some bugs if there is any before the official release of v1beta4.
> 
> It's now hard-coding as false, so that v1beta4 is not possible for `kubeadm init`
> 
> https://github.com/kubernetes/kubernetes/blob/c8125c40299513470c7b992169233edf086755da/cmd/kubeadm/app/util/config/initconfiguration.go#L295
> 
> WDYT?

my vote goes for doing this in the next release cycle 1.30. we can enable it at master for the whole 4 months and see what breaks. i don't want to expose it to the users yet ALA "beta testing" in 1.29.

but if the others +1 your idea let's proceed with it.

chendave commented:
/unhold

pacoxu commented:
/release-note-edit
```
kubeadm: skip image pull for addons that are skipped
```

A better release note may be needed. I just updated it with a simple one.

pacoxu commented:
/release-note-edit
```release-note
kubeadm: skip image pull for addons that are skipped
```

neolit123 commented:
/release-note-edit

```release-note
NONE
```

@pacoxu it's a v1beta4 feature, so no release notes, until it's released.

pacoxu commented:
> @pacoxu it's a v1beta4 feature, so no release notes, until it's released.

My fault. Got it.
